<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="James Ashmore" />


<title>Clustering</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">allantois-wt</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01-quality-control.html">Quality control</a>
    </li>
    <li>
      <a href="02-normalization.html">Normalization</a>
    </li>
    <li>
      <a href="03-feature-selection.html">Feature selection</a>
    </li>
    <li>
      <a href="04-reduced-dimensions.html">Dimensionality reduction</a>
    </li>
    <li>
      <a href="05-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="06-doublet-detection.html">Doublet detection</a>
    </li>
    <li>
      <a href="07-cell-cycle.html">Cell cycle assignment</a>
    </li>
    <li>
      <a href="08-marker-detection.html">Marker gene detection</a>
    </li>
    <li>
      <a href="09-cell-annotation.html">Cell type annotation</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jma1991/allantois-wt">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Clustering</h1>
<h4 class="author">James Ashmore</h4>
<h4 class="date">18 March, 2020</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-03-18
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>allantois-wt/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.1). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200318code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200318)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200318code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200318)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong6b57056"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> 6b57056 </a>
</p>
</div>
<div id="strongRepositoryversionstrong6b57056" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version 6b57056. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/01-quality-control.Rmd
    Untracked:  analysis/02-normalization.Rmd
    Untracked:  analysis/03-feature-selection.Rmd
    Untracked:  analysis/04-reduced-dimensions.Rmd
    Untracked:  analysis/05-clustering.Rmd
    Untracked:  analysis/06-doublet-detection.Rmd
    Untracked:  analysis/07-cell-cycle.Rmd
    Untracked:  analysis/08-marker-detection.Rmd
    Untracked:  analysis/09-cell-annotation.Rmd
    Untracked:  analysis/10-data-integration.Rmd
    Untracked:  analysis/cache/
    Untracked:  data/01-quality-control.rds
    Untracked:  data/02-normalization.rds
    Untracked:  data/03-feature-selection.rds
    Untracked:  data/04-reduced-dimensions.rds
    Untracked:  data/05-clustering.rds
    Untracked:  data/06-doublet-detection.rds
    Untracked:  data/07-cell-cycle.rds
    Untracked:  data/08-marker-detection.rds
    Untracked:  data/09-cell-annotation.rds
    Untracked:  data/markers.csv
    Untracked:  data/original.rds
    Untracked:  data/quant.rds
    Untracked:  markers.csv

Unstaged changes:
    Modified:   _workflowr.yml
    Modified:   analysis/_site.yml

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="motivation" class="section level2">
<h2>Motivation</h2>
<p>Clustering is an unsupervised learning procedure that is used in scRNA-seq data analysis to empirically define groups of cells with similar expression profiles. Here, we demonstrate the application of several commonly used clustering methods with our experiment data.</p>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<p>Set chunk options:</p>
<pre class="r"><code>knitr::opts_chunk$set(
  autodep    = TRUE,
  cache      = TRUE,
  cache.path = &quot;cache/05-clustering.Rmd/&quot;,
  dev        = &quot;png&quot;,
  error      = FALSE,
  message    = FALSE,
  warning    = FALSE
)</code></pre>
<p>Load required packages:</p>
<pre class="r"><code>pacman::p_load(
  cluster,
  devtools,
  here,
  igraph,
  pheatmap,
  readr,
  SC3,
  scater,
  scran
)</code></pre>
<p>Read experiment data:</p>
<pre class="r"><code>rds &lt;- here(&quot;data/04-reduced-dimensions.rds&quot;)

sce &lt;- read_rds(rds)</code></pre>
</div>
<div id="graph" class="section level2">
<h2>Graph</h2>
<div id="walktrap" class="section level3 tabset">
<h3>Walktrap</h3>
<p>Perform clustering using rank-based weights followed by Walktrap clustering:</p>
<pre class="r"><code>snn &lt;- buildSNNGraph(sce, type = &quot;rank&quot;, use.dimred = &quot;PCA&quot;)

cls &lt;- cluster_walktrap(snn)$membership

sce$cluster_walktrap &lt;- factor(cls)</code></pre>
<p>Calculate the modularity of each cluster:</p>
<pre class="r"><code>mod &lt;- clusterModularity(snn, sce$cluster_walktrap, as.ratio = TRUE)

mod &lt;- Matrix::forceSymmetric(mod)

pal &lt;- RColorBrewer::brewer.pal(5, &quot;Greens&quot;)

col &lt;- colorRampPalette(pal)(10)

brk &lt;- seq(0, max(mod), length.out = 11)

pheatmap(mod, color = col, breaks = brk, cluster_rows = FALSE, cluster_cols = FALSE)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Inspect dimensionality reduction plots coloured by explanatory variables:</p>
<div id="pca" class="section level4">
<h4>PCA</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_walktrap&quot;)

plt &lt;- lapply(var, function(x) plotPCA(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-5-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="tsne" class="section level4">
<h4>TSNE</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_walktrap&quot;)

plt &lt;- lapply(var, function(x) plotTSNE(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-6-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="umap" class="section level4">
<h4>UMAP</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_walktrap&quot;)

plt &lt;- lapply(var, function(x) plotUMAP(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="louvain" class="section level3 tabset">
<h3>Louvain</h3>
<p>Perform clustering using Jaccard-based weights followed by Louvain clustering:</p>
<pre class="r"><code>snn &lt;- buildSNNGraph(sce, type = &quot;jaccard&quot;, use.dimred = &quot;PCA&quot;)

cls &lt;- cluster_louvain(snn)$membership

sce$cluster_louvain &lt;- factor(cls)</code></pre>
<p>Calculate the modularity of each cluster:</p>
<pre class="r"><code>mod &lt;- clusterModularity(snn, sce$cluster_louvain, as.ratio = TRUE)

mod &lt;- Matrix::forceSymmetric(mod)

pal &lt;- RColorBrewer::brewer.pal(5, &quot;Greens&quot;)

col &lt;- colorRampPalette(pal)(10)

brk &lt;- seq(0, max(mod), length.out = 11)

pheatmap(mod, color = col, breaks = brk, cluster_rows = FALSE, cluster_cols = FALSE)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Inspect dimensionality reduction plots coloured by explanatory variables:</p>
<div id="pca-1" class="section level4">
<h4>PCA</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_louvain&quot;)

plt &lt;- lapply(var, function(x) plotPCA(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="tsne-1" class="section level4">
<h4>TSNE</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_louvain&quot;)

plt &lt;- lapply(var, function(x) plotTSNE(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="umap-1" class="section level4">
<h4>UMAP</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_louvain&quot;)

plt &lt;- lapply(var, function(x) plotUMAP(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-12-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="supervised" class="section level2">
<h2>Supervised</h2>
<div id="k-means" class="section level3 tabset">
<h3>K-means</h3>
<p>Perform k-means clustering on the PCA matrix:</p>
<pre class="r"><code>set.seed(1701)

dim &lt;- reducedDim(sce, &quot;PCA&quot;)

gap &lt;- clusGap(dim, kmeans, K.max = 50)

num &lt;- maxSE(gap$Tab[, &quot;gap&quot;], gap$Tab[, &quot;SE.sim&quot;])

dat &lt;- data.frame(var = seq_len(50), val = gap$Tab[, &quot;gap&quot;])

ggplot(dat, aes(var, val)) +
  geom_point() +
  geom_vline(xintercept = num, colour = &quot;red&quot;) +
  labs(x = &quot;Number of clusters&quot;, y = &quot;Gap statistic&quot;) +
  theme_classic()</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Select optimal number of clusters:</p>
<pre class="r"><code>cls &lt;- kmeans(dim, centers = num)

sce$cluster_kmeans &lt;- factor(cls$cluster)</code></pre>
<p>Plot the distances between clusters using their centroids:</p>
<pre class="r"><code>tab &lt;- tabulate(cls$cluster)

dat &lt;- data.frame(wcss = cls$withinss, ncells = tab)

dat$rms &lt;- sqrt(dat$wcss / dat$ncells)

hcl &lt;- hclust(stats::dist(cls$centers), method = &quot;ward.D2&quot;)

plot(hcl)</code></pre>
<p>Inspect dimensionality reduction plots coloured by explanatory variables:</p>
<div id="pca-2" class="section level4">
<h4>PCA</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_kmeans&quot;)

plt &lt;- lapply(var, function(x) plotPCA(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-16-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="tsne-2" class="section level4">
<h4>TSNE</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_kmeans&quot;)

plt &lt;- lapply(var, function(x) plotTSNE(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-17-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="umap-2" class="section level4">
<h4>UMAP</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_kmeans&quot;)

plt &lt;- lapply(var, function(x) plotUMAP(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-18-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="unsupervised" class="section level2">
<h2>Unsupervised</h2>
<div id="hierarchical" class="section level3 tabset">
<h3>Hierarchical</h3>
<p>Perform hierarchical cluster analysis on the PCA data:</p>
<pre class="r"><code>dim &lt;- reducedDim(sce, &quot;PCA&quot;)

dst &lt;- dist(dim)

cls &lt;- hclust(dst, &quot;ward.D2&quot;)

mat &lt;- as.matrix(dst)

cut &lt;- dynamicTreeCut::cutreeDynamic(cls, distM = mat)</code></pre>
<pre><code> ..cutHeight not given, setting it to 242  ===&gt;  99% of the (truncated) height range in dendro.
 ..done.</code></pre>
<pre class="r"><code>sce$cluster_hclust &lt;- factor(cut)</code></pre>
<p>Calculate the modularity of each cluster:</p>
<pre class="r"><code>sil &lt;- silhouette(cut, dist = dst)

plot(sil)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Inspect dimensionality reduction plots coloured by explanatory variables:</p>
<div id="pca-3" class="section level4">
<h4>PCA</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_hclust&quot;)

plt &lt;- lapply(var, function(x) plotPCA(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-21-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="tsne-3" class="section level4">
<h4>TSNE</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_hclust&quot;)

plt &lt;- lapply(var, function(x) plotTSNE(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-22-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="umap-3" class="section level4">
<h4>UMAP</h4>
<pre class="r"><code>var &lt;- c(&quot;cluster_hclust&quot;)

plt &lt;- lapply(var, function(x) plotUMAP(sce, colour_by = x, text_by = x))

patchwork::wrap_plots(plt)</code></pre>
<p><img src="figure/05-clustering.Rmd/unnamed-chunk-23-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="selection" class="section level2">
<h2>Selection</h2>
<p>Choose clustering method for downstream analysis:</p>
<pre class="r"><code>sce$cluster &lt;- sce$cluster_louvain</code></pre>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<div id="output" class="section level3">
<h3>Output</h3>
<p>Write experiment data:</p>
<pre class="r"><code>write_rds(sce, here(&quot;data/05-clustering.rds&quot;))</code></pre>
</div>
<div id="session" class="section level3">
<h3>Session</h3>
<p>Print session information:</p>
<pre class="r"><code>session_info()</code></pre>
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.1 (2019-07-05)
 os       macOS Catalina 10.15.3      
 system   x86_64, darwin15.6.0        
 ui       X11                         
 language (EN)                        
 collate  en_GB.UTF-8                 
 ctype    en_GB.UTF-8                 
 tz       Europe/London               
 date     2020-03-18                  

─ Packages ───────────────────────────────────────────────────────────────────
 package              * version  date       lib source           
 assertthat             0.2.1    2019-03-21 [1] CRAN (R 3.6.0)   
 backports              1.1.5    2019-10-02 [1] CRAN (R 3.6.1)   
 beeswarm               0.2.3    2016-04-25 [1] CRAN (R 3.6.1)   
 Biobase              * 2.46.0   2019-10-29 [1] Bioconductor     
 BiocGenerics         * 0.32.0   2019-10-29 [1] Bioconductor     
 BiocNeighbors          1.4.2    2020-02-29 [1] Bioconductor     
 BiocParallel         * 1.20.1   2019-12-21 [1] Bioconductor     
 BiocSingular           1.2.2    2020-02-14 [1] Bioconductor     
 bitops                 1.0-6    2013-08-17 [1] CRAN (R 3.6.0)   
 callr                  3.4.2    2020-02-12 [1] CRAN (R 3.6.1)   
 caTools                1.18.0   2020-01-17 [1] CRAN (R 3.6.1)   
 class                  7.3-15   2019-01-01 [1] CRAN (R 3.6.1)   
 cli                    2.0.2    2020-02-28 [1] CRAN (R 3.6.0)   
 cluster              * 2.1.0    2019-06-19 [1] CRAN (R 3.6.1)   
 codetools              0.2-16   2018-12-24 [1] CRAN (R 3.6.1)   
 colorspace             1.4-1    2019-03-18 [1] CRAN (R 3.6.0)   
 cowplot                1.0.0    2019-07-11 [1] CRAN (R 3.6.0)   
 crayon                 1.3.4    2017-09-16 [1] CRAN (R 3.6.0)   
 DelayedArray         * 0.12.2   2020-01-06 [1] Bioconductor     
 DelayedMatrixStats     1.8.0    2019-10-29 [1] Bioconductor     
 DEoptimR               1.0-8    2016-11-19 [1] CRAN (R 3.6.1)   
 desc                   1.2.0    2018-05-01 [1] CRAN (R 3.6.1)   
 devtools             * 2.2.2    2020-02-17 [1] CRAN (R 3.6.0)   
 digest                 0.6.25   2020-02-23 [1] CRAN (R 3.6.0)   
 doParallel             1.0.15   2019-08-02 [1] CRAN (R 3.6.1)   
 doRNG                  1.8.2    2020-01-27 [1] CRAN (R 3.6.1)   
 dplyr                  0.8.5    2020-03-07 [1] CRAN (R 3.6.0)   
 dqrng                  0.2.1    2019-05-17 [1] CRAN (R 3.6.1)   
 dynamicTreeCut         1.63-1   2016-03-11 [1] CRAN (R 3.6.1)   
 e1071                  1.7-3    2019-11-26 [1] CRAN (R 3.6.1)   
 edgeR                  3.28.1   2020-02-26 [1] Bioconductor     
 ellipsis               0.3.0    2019-09-20 [1] CRAN (R 3.6.0)   
 evaluate               0.14.1   2020-02-14 [1] local            
 fansi                  0.4.1    2020-01-08 [1] CRAN (R 3.6.0)   
 farver                 2.0.3    2020-01-16 [1] CRAN (R 3.6.0)   
 fastmap                1.0.1    2019-10-08 [1] CRAN (R 3.6.0)   
 foreach                1.5.1    2018-11-26 [1] R-Forge (R 3.6.1)
 fs                     1.3.2    2020-03-05 [1] CRAN (R 3.6.0)   
 gdata                  2.18.0   2017-06-06 [1] CRAN (R 3.6.1)   
 GenomeInfoDb         * 1.22.0   2019-10-29 [1] Bioconductor     
 GenomeInfoDbData       1.2.2    2020-02-14 [1] Bioconductor     
 GenomicRanges        * 1.38.0   2019-10-29 [1] Bioconductor     
 ggbeeswarm             0.6.0    2017-08-07 [1] CRAN (R 3.6.1)   
 ggplot2              * 3.3.0    2020-03-05 [1] CRAN (R 3.6.0)   
 git2r                  0.26.1   2019-06-29 [1] CRAN (R 3.6.1)   
 glue                   1.3.1    2019-03-12 [1] CRAN (R 3.6.0)   
 gplots                 3.0.3    2016-08-12 [1] R-Forge (R 3.6.1)
 gridExtra              2.3      2017-09-09 [1] CRAN (R 3.6.1)   
 gtable                 0.3.0    2019-03-25 [1] CRAN (R 3.6.0)   
 gtools                 3.8.1    2018-06-26 [1] CRAN (R 3.6.1)   
 here                 * 0.1      2017-05-28 [1] CRAN (R 3.6.1)   
 hms                    0.5.3    2020-01-08 [1] CRAN (R 3.6.1)   
 htmltools              0.4.0    2019-10-04 [1] CRAN (R 3.6.0)   
 httpuv                 1.5.2    2019-09-11 [1] CRAN (R 3.6.0)   
 igraph               * 1.2.4.2  2019-11-27 [1] CRAN (R 3.6.0)   
 IRanges              * 2.20.2   2020-01-13 [1] Bioconductor     
 irlba                  2.3.3    2019-02-05 [1] CRAN (R 3.6.1)   
 iterators              1.0.12   2019-07-26 [1] CRAN (R 3.6.1)   
 KernSmooth             2.23-16  2019-10-15 [1] CRAN (R 3.6.1)   
 knitr                  1.28     2020-02-06 [1] CRAN (R 3.6.0)   
 labeling               0.3      2014-08-23 [1] CRAN (R 3.6.0)   
 later                  1.0.0    2019-10-04 [1] CRAN (R 3.6.0)   
 lattice                0.20-40  2020-02-19 [1] CRAN (R 3.6.0)   
 lifecycle              0.2.0    2020-03-06 [1] CRAN (R 3.6.0)   
 limma                  3.42.2   2020-02-03 [1] Bioconductor     
 locfit                 1.5-9.1  2013-04-20 [1] CRAN (R 3.6.1)   
 magrittr               1.5      2014-11-22 [1] CRAN (R 3.6.0)   
 Matrix                 1.2-18   2019-11-27 [1] CRAN (R 3.6.0)   
 matrixStats          * 0.55.0   2019-09-07 [1] CRAN (R 3.6.0)   
 memoise                1.1.0    2017-04-21 [1] CRAN (R 3.6.0)   
 mime                   0.9      2020-02-04 [1] CRAN (R 3.6.0)   
 munsell                0.5.0    2018-06-12 [1] CRAN (R 3.6.0)   
 mvtnorm                1.1-0    2020-02-07 [1] R-Forge (R 3.6.1)
 pacman                 0.5.1    2019-03-11 [1] CRAN (R 3.6.1)   
 patchwork              1.0.0    2019-12-01 [1] CRAN (R 3.6.0)   
 pcaPP                  1.9-73   2018-01-14 [1] CRAN (R 3.6.1)   
 pheatmap             * 1.0.12   2019-01-04 [1] CRAN (R 3.6.1)   
 pillar                 1.4.3    2019-12-20 [1] CRAN (R 3.6.0)   
 pkgbuild               1.0.6    2019-10-09 [1] CRAN (R 3.6.1)   
 pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 3.6.0)   
 pkgload                1.0.2    2018-10-29 [1] CRAN (R 3.6.1)   
 prettyunits            1.1.1    2020-01-24 [1] CRAN (R 3.6.1)   
 processx               3.4.2    2020-02-09 [1] CRAN (R 3.6.0)   
 promises               1.1.0    2019-10-04 [1] CRAN (R 3.6.0)   
 ps                     1.3.2    2020-02-13 [1] CRAN (R 3.6.1)   
 purrr                  0.3.3    2019-10-18 [1] CRAN (R 3.6.0)   
 R6                     2.4.1    2019-11-12 [1] CRAN (R 3.6.0)   
 RColorBrewer           1.1-2    2014-12-07 [1] CRAN (R 3.6.0)   
 Rcpp                   1.0.3    2019-11-08 [1] CRAN (R 3.6.0)   
 RCurl                  1.98-1.1 2020-01-19 [1] CRAN (R 3.6.0)   
 readr                * 1.3.1    2018-12-21 [1] CRAN (R 3.6.1)   
 remotes                2.1.1    2020-02-15 [1] CRAN (R 3.6.1)   
 rlang                  0.4.5    2020-03-01 [1] CRAN (R 3.6.0)   
 rmarkdown              2.1      2020-01-20 [1] CRAN (R 3.6.1)   
 rngtools               1.5      2020-01-23 [1] CRAN (R 3.6.1)   
 robustbase             0.93-6   2019-12-12 [1] R-Forge (R 3.6.1)
 ROCR                   1.0-7    2015-03-26 [1] CRAN (R 3.6.1)   
 rprojroot              1.3-2    2018-01-03 [1] CRAN (R 3.6.1)   
 rrcov                  1.5-2    2020-01-16 [1] CRAN (R 3.6.1)   
 rsvd                   1.0.3    2020-02-17 [1] CRAN (R 3.6.0)   
 S4Vectors            * 0.24.3   2020-01-18 [1] Bioconductor     
 SC3                  * 1.14.0   2019-10-29 [1] Bioconductor     
 scales                 1.1.0    2019-11-18 [1] CRAN (R 3.6.0)   
 scater               * 1.14.6   2019-12-16 [1] Bioconductor     
 scran                * 1.14.6   2020-02-03 [1] Bioconductor     
 sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 3.6.1)   
 shiny                  1.4.0.1  2020-03-12 [1] CRAN (R 3.6.1)   
 SingleCellExperiment * 1.8.0    2019-10-29 [1] Bioconductor     
 statmod                1.4.34   2020-02-17 [1] CRAN (R 3.6.0)   
 stringi                1.4.6    2020-02-17 [1] CRAN (R 3.6.0)   
 stringr                1.4.0    2019-02-10 [1] CRAN (R 3.6.0)   
 SummarizedExperiment * 1.16.1   2019-12-19 [1] Bioconductor     
 testthat               2.3.2    2020-03-02 [1] CRAN (R 3.6.0)   
 tibble                 2.1.3    2019-06-06 [1] CRAN (R 3.6.0)   
 tidyselect             1.0.0    2020-01-27 [1] CRAN (R 3.6.0)   
 usethis              * 1.5.1    2019-07-04 [1] CRAN (R 3.6.1)   
 vctrs                  0.2.4    2020-03-10 [1] CRAN (R 3.6.0)   
 vipor                  0.4.5    2017-03-22 [1] CRAN (R 3.6.0)   
 viridis                0.5.1    2018-03-29 [1] CRAN (R 3.6.1)   
 viridisLite            0.3.0    2018-02-01 [1] CRAN (R 3.6.0)   
 withr                  2.1.2    2018-03-15 [1] CRAN (R 3.6.0)   
 workflowr            * 1.6.1    2020-03-11 [1] CRAN (R 3.6.1)   
 WriteXLS               5.0.0    2019-05-25 [1] CRAN (R 3.6.1)   
 xfun                   0.12     2020-01-13 [1] CRAN (R 3.6.1)   
 xtable                 1.8-4    2019-04-21 [1] CRAN (R 3.6.0)   
 XVector                0.26.0   2019-10-29 [1] Bioconductor     
 yaml                   2.2.1    2020-02-01 [1] CRAN (R 3.6.0)   
 zlibbioc               1.32.0   2019-10-29 [1] Bioconductor     

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
